nixos/modules/virtualisation/containers-next/default.nix

```diff
@@ -0,0 +1,571 @@
+        message = ''
+          No configuration found for zone `${inst.zone}'!
+          (Invalid container: ${n})
+        '';
+      }
+      { assertion = inst.zone != null -> dynamicAddrsDisabled inst;
+        message = ''
+          Cannot assign additional generic address-pool to a veth-pair if corresponding
+          container `${n}' already uses zone `${inst.zone}'!
+        '';
+      }
+      { assertion = !inst.sharedNix -> inst.activation.strategy != "reload";
+        message = ''
+          Cannot reload a container with `sharedNix' disabled! As soon as the
+          `BindReadOnly='-options change, a config activation can't be done without a reboot
+          (affected: ${n})!
+        '';
+      }
+    ]));
+
+    services.radvd = {
+      inherit (radvd) config enable;
+    };
+
+    nixos.containers.rendered = mapAttrs (const (x: x.container)) images;
+
+    systemd = {
+      network.networks = mkMerge
+        ((flip mapAttrsToList cfg (name: config: optionalAttrs (config.network != null && config.zone == null) {
+          "20-${ifacePrefix "veth"}-${name}" = {
+            matchConfig = mkMatchCfg "veth" name;
+            address = config.network.v4.addrPool
+              ++ config.network.v6.addrPool
+              ++ optionals (config.network.v4.static.hostAddresses != null)
+                config.network.v4.static.hostAddresses
+              ++ optionals (config.network.v6.static.hostAddresses != null)
+                config.network.v6.static.hostAddresses;
+            networkConfig = mkNetworkCfg (config.network.v4.addrPool != []) config.network.v4.nat;
+          };
+        }))
+        ++ (flip mapAttrsToList config.nixos.containers.zones (name: zone: {
+          "20-${ifacePrefix "zone"}-${name}" = {
+            matchConfig = mkMatchCfg "zone" name;
+            address = zone.v4.addrPool
+              ++ zone.v6.addrPool
+              ++ zone.hostAddresses
+              ++ (flatten (flip mapAttrsToList cfg
+                (name: config: optionals (config.zone != null && config.network != null)
+                  (config.network.v4.static.hostAddresses ++ config.network.v6.static.hostAddresses)
+                )));
+            networkConfig = mkNetworkCfg true zone.v4.nat;
+          };
+        })));
+
+      nspawn = mapAttrs (const mkContainer) images;
+      targets.machines.wants = map (x: "systemd-nspawn@${x}.service") (attrNames cfg);
+      services = listToAttrs (flip map (attrNames cfg) (container:
+        let
+          inherit (cfg.${container}) activation credentials;
+        in nameValuePair "systemd-nspawn@${container}" {
+          preStart = mkBefore ''
+            if [ ! -d /var/lib/machines/${container} ]; then
+              mkdir -p /var/lib/machines/${container}/{etc,var,nix/var/nix}
+              touch /var/lib/machines/${container}/etc/{os-release,machine-id} || true
```

Comment:
I wonder: why should those fail?

Reply:
It's a symlink to a store-path as soon as `/init` is executed the first time.

Reply:
OK, actually this shouldn't be relevant anymore because of the `! -d` check above (the `|| true` was added before that if existed). Fixed.